<a href="https://colab.research.google.com/github/MinakoNG63/DSFB/blob/main/18_Classification_with_scikit_learn_63070240.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 18) Basic Classification with Scitkit-learn

Term 1 2022 - Instructor: Teerapong Leelanupab

Teaching Assistant:
1. Piyawat Chuangkrud (Sam)
2. Suvapat Manu (Mint)

***

In [ ]:
!mkdir data

# download diabetes.csv
!gdown --id 1lQERpvrkfR_SQzygWssdsUuAfHEIHYsy -O data/diabetes.csv

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1lQERpvrkfR_SQzygWssdsUuAfHEIHYsy
To: /content/data/diabetes.csv
100% 27.4k/27.4k [00:00<00:00, 32.8MB/s]


Classification involves using labeled (known) training examples to make an accurate prediction for new unseen input examples. In this lab we will use the classification functionality provided by the Scitkit-learn Python package.

The *k-Nearest Neighbour (KNN) classifier* is a simple but effective "lazy"classifier. Given a new input example, it finds the most similar previous examples for which a decision has already been made (i.e. their nearest neighbours from the training set). Based on the majority vote among the K neighbours, a prediction will be made for the input.

#### Example 1: KNN Classifier

The scikit-learn package includes a number of datasets, which are useful for getting a handle on a given machine learning algorithm before using it in your own work. We will load the version of the Iris dataset which is provided by scikit-learn:

In [ ]:
from sklearn.datasets import load_iris
iris = load_iris()

In [ ]:
type(iris)

sklearn.utils.Bunch

This dataset has four different descriptive features:

In [ ]:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


Each example in the dataset has a class label or a "target" from three possible classes:

In [ ]:
print(iris.target_names)

['setosa' 'versicolor' 'virginica']


In [ ]:
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [ ]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Build a nearest neighbour classifier using $k=1$ nearest neighbour. In this case we will use the full dataset and all of the target labels for those examples:

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1)
knn.fit(iris.data, iris.target)

KNeighborsClassifier(n_neighbors=1)

We can test it out by making a prediction for a new input example described by 4 feature values:

In [ ]:
import numpy as np
xinput = np.array([[3.0, 5.0, 4.1, 2.0]])
# make the prediction, the output is the number of the class
pred_class_number = knn.predict(xinput)
# to get the name of the class
print( iris.target_names[pred_class_number] )

['virginica']


We can also predict for multiple input examples at once:

In [ ]:
xinput = np.array([[3, 5, 4, 2], [3, 5, 2, 2]])
pred_class_numbers = knn.predict(xinput)
print( iris.target_names[pred_class_numbers] )

['virginica' 'setosa']


#### Example 2: KNN Classifier

Use the same data and convert them into the dataframe like a common way we load data from csv, which contains both features and class labels in the same file

In [ ]:
# This is just to see our data using Pandas Data Frame and because our features and lable are already separated in two arrays. So, we don't need to do anything.
import pandas as pd
features = pd.DataFrame(iris.data, columns = ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)'])
features

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

In [ ]:
labels = pd.DataFrame(iris.target, columns = ['iris class'])
labels

iris class
0             0
1             0
2             0
3             0
4             0
..          ...
145           2
146           2
147           2
148           2
149           2

[150 rows x 1 columns]

In [ ]:
iris_df = pd.concat([features, labels], axis=1)
iris_df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     iris class  
0             0  
1             0  
2             0  
3             0  
4             0  
..          ...  
145           2  
146           2  
147           2  
148           2  
149           2  

[150 rows x 5 columns]

In [ ]:
iris_df.shape

(150, 5)

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split

def simple_hold_out_spliting(df, label_column, train_size):
    # Add code here
    #Isolate the descriptive features we are interested in
    X = df.drop(label_column, axis=1)
    Y = df[label_column]

    X_train, X_test, y_train, y_test \
        = train_test_split(X, Y, random_state=0, \
                                    train_size = train_size)
    return X_train, X_test, y_train, y_test

In [ ]:
iris_train, iris_test, target_train, target_test = simple_hold_out_spliting(iris_df, "iris class", 0.7)

In [ ]:
model = KNeighborsClassifier(n_neighbors=3)
model.fit(iris_train, target_train)
print(model)

KNeighborsClassifier(n_neighbors=3)


In [ ]:
predicted = model.predict(iris_test)
predicted

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 2, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0, 2, 1, 0, 2, 2, 1, 0, 2, 1, 1, 2, 0, 2, 0,
       0])

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(target_test, predicted)

0.9777777777777777

#### Example 3: KNN Classifier

Next, we will load a CSV copy of the Pima Indian diabetes dataset from the UCI Machine Learning Repository, where the target is to make a prediction of 1 (tested positive for diabetes) or 0 (tested negative for diabetes).

In [ ]:
import pandas as pd
# load the CSV file as a numpy matrix
df = pd.read_csv("data/diabetes.csv")
print(df.shape)
df.head(10)

(739, 9)


preg  plas  pres  skin  insu  mass   pedi  age     class
0     6   148    72    35     0  33.6  0.627   50  positive
1     1    85    66    29     0  26.6  0.351   31  negative
2     8   183    64     0     0  23.3  0.672   32  positive
3     1    89    66    23    94  28.1  0.167   21  negative
4     0   137    40    35   168  43.1  2.288   33  positive
5     5   116    74     0     0  25.6  0.201   30  negative
6     3    78    50    32    88  31.0  0.248   26  positive
7    10   115     0     0     0  35.3  0.134   29  negative
8     2   197    70    45   543  30.5  0.158   53  positive
9     8   125    96     0     0   0.0  0.232   54  positive

The CSV format of the dataset contains data for 739 rows (patients), each with 9 columns. These are 8 descriptive numeric features, and the binary target value. We will separate out the descriptive columns from the target column (i.e. the class labels).

In [ ]:
#convert from dataframe to numpy array
raw_dataset = df.values
#seperate data into descriptive features and a target class
dataset = raw_dataset[:,0:7]
target = raw_dataset[:,8]

In [ ]:
print(dataset)

[[6 148 72 ... 0 33.6 0.627]
 [1 85 66 ... 0 26.6 0.351]
 [8 183 64 ... 0 23.3 0.672]
 ...
 [0 126 86 ... 120 27.4 0.515]
 [8 65 72 ... 0 32.0 0.6]
 [2 99 60 ... 160 36.6 0.453]]


In [ ]:
print(target)

['positive' 'negative' 'positive' 'negative' 'positive' 'negative'
 'positive' 'negative' 'positive' 'positive' 'negative' 'positive'
 'negative' 'positive' 'positive' 'positive' 'positive' 'positive'
 'negative' 'positive' 'negative' 'negative' 'positive' 'positive'
 'positive' 'positive' 'positive' 'negative' 'negative' 'negative'
 'negative' 'positive' 'negative' 'negative' 'negative' 'negative'
 'negative' 'positive' 'positive' 'positive' 'negative' 'negative'
 'negative' 'positive' 'negative' 'positive' 'negative' 'negative'
 'positive' 'negative' 'negative' 'negative' 'negative' 'positive'
 'negative' 'negative' 'positive' 'negative' 'negative' 'negative'
 'negative' 'positive' 'negative' 'negative' 'positive' 'negative'
 'positive' 'negative' 'negative' 'negative' 'positive' 'negative'
 'positive' 'negative' 'negative' 'negative' 'negative' 'negative'
 'positive' 'negative' 'negative' 'negative' 'negative' 'negative'
 'positive' 'negative' 'negative' 'negative' 'positive' 'negat

Now, we will randomly split the complete dataset into a training test (used to build the model) and an unseen test set (used to try out and evaluate the model). Scikit-learn provides functionality to do this. We will specify that 20% (0.2) of the data will be used for the test set.

In [ ]:
from sklearn.model_selection import train_test_split
dataset_train, dataset_test, target_train, target_test = train_test_split(dataset, target, test_size=0.2)

In [ ]:
print("Training set size is %d" % dataset_train.shape[0] )
print("Test set size is %d" % dataset_test.shape[0] )

Training set size is 591
Test set size is 148


Next, we will fit a k-nearest neighbor model to the data using $k=3$ nearest neighbours:

In [ ]:
model = KNeighborsClassifier(n_neighbors=3)
model.fit(dataset_train, target_train)
print(model)

KNeighborsClassifier(n_neighbors=3)


Make predictions for the test set, based on the model that we just built:

In [ ]:
predicted = model.predict(dataset_test)
predicted

array(['positive', 'positive', 'negative', 'negative', 'negative',
       'positive', 'negative', 'positive', 'negative', 'negative',
       'negative', 'negative', 'positive', 'positive', 'positive',
       'negative', 'negative', 'negative', 'positive', 'negative',
       'negative', 'negative', 'positive', 'negative', 'positive',
       'negative', 'positive', 'negative', 'negative', 'positive',
       'negative', 'negative', 'negative', 'positive', 'negative',
       'negative', 'negative', 'negative', 'negative', 'positive',
       'negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'negative', 'positive', 'negative', 'negative',
       'positive', 'positive', 'positive', 'negative', 'negative',
       'positive', 'negative', 'negative', 'positive', 'negative',
       'negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'positive', 'positive', 'positive', 'negative',
       'positive', 'positive', 'negative', 'negative', 'negati

In [ ]:
num_pos = (predicted == 'positive').sum()
num_neg = (predicted == 'negative').sum()
print( "Number of patients predicted positive for diabetes: %d" % num_pos )
print( "Number of patients predicted negative for diabetes: %d" % num_neg )

Number of patients predicted positive for diabetes: 45
Number of patients predicted negative for diabetes: 103


We can compare our predictions to the "correct answer" based on the labels for the test data:

In [ ]:
print("Predictions\n", predicted)
print("Correct labels\n", target_test)

Predictions
 [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2 1 1 2 0 2 0 0]
Correct labels
 114    2
62     1
33     0
107    2
7      0
100    2
40     0
86     1
76     1
71     1
134    2
51     1
73     1
54     1
63     1
37     0
78     1
90     1
45     0
16     0
121    2
66     1
24     0
8      0
126    2
22     0
44     0
97     1
93     1
26     0
137    2
84     1
27     0
127    2
132    2
59     1
18     0
83     1
61     1
92     1
112    2
2      0
141    2
43     0
10     0
Name: iris class, dtype: int64


We can quantitatively check how accurate these predictions are, by measuring *accuracy*, which will return a value between 0.0 (predictions are completely wrong) and 1.0 (predictions are 100% accurate):

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(target_test, predicted)

0.9777777777777777

In the laterlab we will look at evaluation measures for classification in more detail.